File: framework/languages/node/samples/SupportChat.Ts/Server/Support/Adapters/ZLink/Handlers/allocate-conversation-handler.ts
import { Inject } from '@nestjs/common';
import { zlinkRequestHandler } from '@zlink-systems/nestjs';
import { SupportConversationAllocator } from '../../../Application/ConversationAssignment/support-conversation-allocator';
import { ConversationStatuses, PacketNames, allocateConversationRes } from '../../../../../Shared/Contracts/messages';
import type { ZLinkRequestHandler } from '@zlink-systems/framework';
import type { SupportConversationAllocator as SupportConversationAllocatorType } from '../../../Application/ConversationAssignment/support-conversation-allocator';
import type {
  AllocateConversationReq,
  AllocateConversationRes
} from '../../../../../Shared/Contracts/messages';

@zlinkRequestHandler('support', PacketNames.allocateConversationReq)
class AllocateConversationHandler implements ZLinkRequestHandler<AllocateConversationReq, AllocateConversationRes> {
  constructor(@Inject(SupportConversationAllocator) private readonly allocator: SupportConversationAllocatorType) {}

  async handle(request: AllocateConversationReq): Promise<AllocateConversationRes> {
    const conversationId = await this.allocator.allocate(
      request.customerActorId,
      request.customerDisplayName,
      request.subject
    );
    return allocateConversationRes(conversationId, ConversationStatuses.WaitingForAgent);
  }
}

export { AllocateConversationHandler };

File: framework/languages/dotnet/samples/SupportChat/Server/Support/Infrastructure/ZLink/Handlers/AllocateConversationHandler.cs
using SupportChat.Server.Configuration;
using SupportChat.Server.Support.Application.ConversationAssignment;
using SupportChat.Shared.Contracts;
using Microsoft.Extensions.Logging;
using Zlink.Framework.Contracts.Handlers;

namespace SupportChat.Server.Support.Infrastructure.ZLink.Handlers;

[ZLinkHandlerGroup("support")]
internal sealed class AllocateConversationHandler(
    SupportConversationAllocator allocator,
    ILogger<AllocateConversationHandler> logger)
    : IZLinkRequestHandler<AllocateConversationReq, AllocateConversationRes>
{
    public async ValueTask<AllocateConversationRes> HandleAsync(
        AllocateConversationReq request,
        ZLinkRequestContext context,
        CancellationToken cancellationToken)
    {
        logger.LogInformation(
            "support allocate: request customer={CustomerActorId} subject={Subject}",
            request.CustomerActorId,
            request.Subject);
        var conversationId = await allocator.AllocateAsync(
            request.CustomerActorId,
            request.CustomerDisplayName,
            request.Subject,
            cancellationToken);
        logger.LogInformation("support allocate: created conversation={ConversationId}", conversationId);
        return new AllocateConversationRes(
            conversationId,
            ConversationStatuses.WaitingForAgent);
    }
}
